app_history.d.ts

```diff
@@ -0,0 +1,98 @@
+interface Window {
+  readonly appHistory: AppHistory;
+}
+
+declare class AppHistory extends EventTarget {
+  readonly current: AppHistoryEntry|null;
+  readonly transition: AppHistoryTransition|null;
+
+  entries(): AppHistoryEntry[];
+
+  readonly canGoBack: boolean;
+  readonly canGoForward: boolean;
+
+  navigate(url: string, options?: AppHistoryNavigateOptions): Promise<void>;
+  reload(options?: AppHistoryReloadOptions): Promise<void>;
+
+  goTo(key: string, options?: AppHistoryNavigationOptions): Promise<void>;
+  back(options?: AppHistoryNavigationOptions): Promise<void>;
+  forward(options?: AppHistoryNavigationOptions): Promise<void>;
+
+  onnavigate: ((this: AppHistory, ev: Event) => any)|null;
+  onnavigatesuccess: ((this: AppHistory, ev: Event) => any)|null;
+  onnavigateerror: ((this: AppHistory, ev: Event) => any)|null;
+}
+
+declare class AppHistoryTransition {
+  readonly navigationType: AppHistoryNavigationType;
+  readonly from: AppHistoryEntry;
+  readonly finished: Promise<void>;
+
+  rollback(options?: AppHistoryNavigationOptions): Promise<void>;
+}
+
+declare class AppHistoryEntry extends EventTarget {
+  readonly key: string;
+  readonly id: string;
+  readonly url: string;
+  readonly index: number;
+  readonly sameDocument: boolean;
+
+  getState(): unknown;
+
+  onnavigateto: ((this: AppHistoryEntry, ev: Event) => any)|null;
+  onnavigatefrom: ((this: AppHistoryEntry, ev: Event) => any)|null;
+  onfinish: ((this: AppHistoryEntry, ev: Event) => any)|null;
+  ondispose: ((this: AppHistoryEntry, ev: Event) => any)|null;
+}
+
+type AppHistoryNavigationType = 'reload'|'push'|'replace'|'traverse';
+
+declare class AppHistoryNavigationOptions {
```

Comment:
So, the input dictionaries should stay as `interface`, sorry. This includes `AppHistoryNavigationOptions`, `AppHistoryNavigateOptions`, `AppHistoryReloadOptions`, and `AppHistoryNavigateEventInit`.

(Basically, Web IDL `interface` means TypeScript `class`, and Web IDL `dictionary` means TypeScript `interface`. I blame Web IDL for being confusing!)

Reply:
Whoops got it.